Refactor log driver reader

cpuguy83 opened the pull request:
Split reader interface from logger interface

Implement new reader interface on jsonfile.
Moves jsonlog decoding from daemon to jsonfile logger.

cc @lk4d4

cpuguy83 committed:
Split reader interface from logger interface

Implement new reader interface on jsonfile.
Moves jsonlog decoding from daemon to jsonfile logger.

Signed-off-by: Brian Goff <cpuguy83@gmail.com>

cpuguy83 committed:
vendor fsnotify

Signed-off-by: Brian Goff <cpuguy83@gmail.com>

cpuguy83 commented:
And for some reason this is hanging on the slow reader test... even when implemented exactly as it is today, except reading from the log file instead of directly from the container pipes.

cpuguy83 committed:
remove dead code after decoupling from jsonlog

Signed-off-by: Brian Goff <cpuguy83@gmail.com>

cpuguy83 commented:
Figured out issue with the SlowConsumer test... which was actually locking up since there was no way to tell the logdriver to stop reading when the container stops... this is fixed.

Unfortunately on the SlowConsumer test I am always 2 bytes short.

LK4D4 commented:
@cpuguy83 Maybe you need to increase timeout, lol :)

LK4D4 commented on api/server/server.go:
```diff
@@ -624,6 +624,17 @@ func (s *Server) getContainersLogs(version version.Version, w http.ResponseWrite
 		closeNotifier = notifier.CloseNotify()
 	}
 
+	c, err := s.daemon.Get(vars["name"])
```
This actually breaking some convention, which was invented by @duglin and @runcom

cpuguy83 replied:
Yes, the reason I'm doing this is because otherwise we end up returning a 200 even though there is no container.
I think we also did this for the stats endpoint.

runcom replied:
Yes but sometimes this is needed and we can't do anything (like hijacking the request and not being able to return 404, same case as this one I suppose)

LK4D4 commented on daemon/container.go:
```diff
@@ -858,32 +860,26 @@ func (c *Container) Attach(stdin io.ReadCloser, stdout io.Writer, stderr io.Writ
+			return logger.ReadLogsNotSupported
+		}
+		var t time.Time
+		logs, err := cLog.ReadLogs(t, -1, false, false, nil)
```
Maybe struct? 5 params too much for me :)

LK4D4 commented on daemon/logger/jsonfilelog/jsonfilelog.go:
```diff
@@ -39,9 +45,10 @@ func New(ctx logger.Context) (logger.Logger, error) {
+		f:       log,
+		buf:     bytes.NewBuffer(nil),
+		ctx:     ctx,
+		readers: []chan struct{}{},
```
can't you use `pkg/pubsub` for this?

LK4D4 commented:
Looks really nice. But I want to suggest you try to use `pkg/pubsub` package. It can reduce amount of code. Feel free to fix it if needed of course.

cpuguy83 commented:
@LK4D4 I updated to use pubsub for tracking readers that need to be closed out when the container stops, but I think it won't work so well for broadcasting log messages to different readers.

cpuguy83 commented:
Also, still have the issue with the slowconsumer test... it's like we're missing a log line or something, somehow.

LK4D4 commented:
@cpuguy83 yeah, you need to put logs everywhere to find it :)

LK4D4 commented on daemon/logger/jsonfilelog/jsonfilelog.go:
```diff
@@ -39,9 +46,10 @@ func New(ctx logger.Context) (logger.Logger, error) {
+		f:       log,
+		buf:     bytes.NewBuffer(nil),
+		ctx:     ctx,
+		readers: pubsub.NewPublisher(3*time.Second, 0),
```
Okay, I'm not sure. But timer in `pubsub` very slow and heavy, it can be okay I think for now.

LK4D4 commented on daemon/container.go:
```diff
@@ -888,33 +878,35 @@ func (c *Container) Attach(stdin io.ReadCloser, stdout io.Writer, stderr io.Writ
+				logrus.Errorf("Error streaming logs: %v", err)
+				break
 			}
+			break
```
I don't like this flow. And I'm not sure how can you reach this code. So I think it is better to remove this break and continue above.

cpuguy83 replied:
Should I use a `goto` here?

LK4D4 commented on daemon/logger/logger.go:
```diff
@@ -16,10 +20,52 @@ type Message struct {
+// NewLogWatcher returns a new LogWatcher.
+func NewLogWatcher() *LogWatcher {
+	return &LogWatcher{
+		Msg:           make(chan Message, 10),
```
I think 10 is supersmall buffer for this task. Also this should be const.

LK4D4 commented on daemon/logger/logger.go:
```diff
@@ -16,10 +20,52 @@ type Message struct {
+	return &LogWatcher{
+		Msg:           make(chan Message, 10),
+		Err:           make(chan error, 10),
+		closeNotifier: make(chan struct{}, 1),
```
should be just `chan struct{}` without buffer

LK4D4 commented on daemon/logger/logger.go:
```diff
@@ -16,10 +20,52 @@ type Message struct {
+
+func (w *LogWatcher) Close() error {
+	select {
+	case w.closeNotifier <- struct{}{}:
```
I don't understand this code. Is it possible to call Close() twice? I think you should just close channel and be done with it.

LK4D4 commented on daemon/logger/logger.go:
```diff
@@ -16,10 +20,52 @@ type Message struct {
+	}
+}
+
+func (w *LogWatcher) Closed() <-chan struct{} {
```
Pretty weird name as for me.

LK4D4 commented on daemon/logger/logger.go:
```diff
@@ -16,10 +20,52 @@ type Message struct {
+	}
+}
+
+func (w *LogWatcher) Close() error {
```
Also would you mind to document this funcs?

LK4D4 commented on daemon/logger/jsonfilelog/jsonfilelog.go:
```diff
@@ -85,3 +94,137 @@ func (l *JSONFileLogger) Close() error {
+}
+
+func (l *JSONFileLogger) readLogs(logWatcher *logger.LogWatcher, config logger.ReadConfig) {
+	defer func() {
```
Any particular reason for creating closure here?

cpuguy83 replied:
Nope, was doing other stuff, in here, forgot to remove it when I got rid of the other stuff.

LK4D4 commented on daemon/logs.go:
```diff
@@ -28,152 +20,66 @@ type ContainerLogsConfig struct {
-			if stderrPipe != nil {
-				stderrPipe.Close()
-				<-chErrStderr
+		case err, ok := <-logs.Err:
```
you don't need `ok` here. Just use `if err != nil { log }`

LK4D4 commented on daemon/logger/jsonfilelog/jsonfilelog.go:
```diff
@@ -85,3 +94,137 @@ func (l *JSONFileLogger) Close() error {
 	return Name
 }
+
+func decodeLogLine(dec *json.Decoder, l *jsonlog.JSONLog) (msg logger.Message, err error) {
```
You definitely don't need named returns here. And use named literals for creating msg too pls

LK4D4 replied:
Also you need to return pointer of course.

LK4D4 commented on daemon/logger/logger.go:
```diff
@@ -16,10 +20,52 @@ type Message struct {
+// LogWatcher is used when consuming logs read from the LogReader interface
+type LogWatcher struct {
+	// For sending log messages to a reader
+	Msg chan Message
```
`chan *Message` of course :)

LK4D4 commented on pkg/broadcastwriter/broadcastwriter.go:
```diff
@@ -4,30 +4,19 @@ import (
-	buf      *bytes.Buffer
-	jsLogBuf *bytes.Buffer
-	streams  map[string](map[io.WriteCloser]struct{})
+	buf     *bytes.Buffer
```
buf isn't used anymore as I see

LK4D4 commented:
Okay, we have to profile it before release. It is very nice that we cleaned up critical part of docker - broadcastwriter.

jfrazelle commented:
LGTM

tiborvass commented:
@cpuguy83 @LK4D4 how is this a refactor? it's an optimization too now that it's using fsnotify ;)

LK4D4 commented:
@tiborvass haha, it is side-effect :)

calavera commented on daemon/container.go:
```diff
@@ -888,32 +878,33 @@ func (c *Container) Attach(stdin io.ReadCloser, stdout io.Writer, stderr io.Writ
+		}
+		logs := cLog.ReadLogs(logger.ReadConfig{Tail: -1})
+
+	LogLoop:
```
I don't think we need the label, do we? there is only one loop.

cpuguy83 replied:
We need to break from inside the switch.

LK4D4 replied:
@calavera yeah, that's pretty weird stuff, that break and continue works for select :)

calavera replied:
oh that's indeed weird, carry on then!

icecrime commented:
That's great :+1: I haven't reviewed in every details, but the overview looks good.

LK4D4 commented:
@cpuguy83 ooooh, I'm sorry, need rebase.

cpuguy83 commented:
Rebased.

In order to easily support tailing across multiple files, and to remove the jsonfile implementation leakage of passing "latest" to the API,  introduced by the rollover PR, I started to implement a MultiReadSeeker, which basically logically concatenates multiple read seekers together (and hence, multiple *os.File readers).... and then I found http://play.golang.org/p/BNL9QCnUHB which actually works quite well and is currently used unmodified in `pkg/ioutils/multireader.go`
Reverted the changes where "latest" replaced "all" from the afformentioned PR.

cpuguy83 commented:
Actually I found a bug with it while writing some tests for that new read seeker. Working on a fix.

LK4D4 commented:
@cpuguy83 also need another rebase :(

cpuguy83 commented:
Rebased, and all (at least pending CI anyway) seems well.
Ended up writing my own MultiReadSeeker as the one linked above turned out to be a bit too limiting.

LK4D4 commented:
@cpuguy83 vet check failed

cpuguy83 commented:
Fixed vet error.
Janky failure is some weird libnetwork error in the docker-py tests.

LK4D4 commented:
@cpuguy83 Tested with `--log-opt max-size=1k`. After some time `logs -f` stopping and then exiting. Maybe we should have some retry logic or some file lock.

cpuguy83 commented:
@LK4D4 Moved the rotate notification down to the bottom of the rotate logic,  (after the file is re-opened and set in the logger struct), and also added some retry logic to the json decoding.
This seemed to be primarily an issue with super fast log writes from what I saw.

LK4D4 commented:
@cpuguy83 hm :/ I don't know what's going, but I ran three logs followers and they losing messages. Not all at once, but from time to time one or two just stuck for short period and then follow again, but messages are lost.

cpuguy83 commented:
@LK4D4 Really strange.... definitely happening around the rotation. Will have to investigate further.
At what point are you seeing the json error?

cpuguy83 commented:
Updated with a buffer (1) on the rotate notifier.

Running (60B max log, writing a log every 20000 microsec)
`docker run -d --name test --log-opt max-size=60B busybox sh -c 'let i=1; while true; do echo $i; let i=$i+1; usleep 20000; done'`
And then `docker logs -f test | go run main.go` ( main.go showed below )

Seems to take some time before we drop anything at all.
If the writes are less than 20000us and it drops fairly quickly.
Note that 1 log entry > 60B (the max size of the log file).
Without the buffer it starts loosing log lines in no time at all.
I don't think a buffer larger than 1 makes sense (and would probably be worse).

```golang
// main.go
package main

import (
	"bufio"
	"log"
	"os"
	"strconv"
)

func main() {
	scanner := bufio.NewScanner(os.Stdin)

	var i int
	for scanner.Scan() {
		in := scanner.Text()

		n, err := strconv.Atoi(in)
		if err != nil {
			log.Fatal(err)
		}

		if i == 0 {
			i = n
		}

		if n != i {
			log.Fatalf("error: %d", n)
		}

		i += 1
	}
}
```

LK4D4 commented:
Works okay for me.
LGTM

calavera commented:
still LGTM, merging :confetti_ball: :exclamation:
